File: java/set/src/LinkList.java
import java.lang.instrument.IllegalClassFormatException;

public class LinkList<E> {
    private class Node {
        public E e;
        public Node next;

        public Node(E e, Node next) {
            this.e = e;
            this.next = next;
        }

        public Node(E e) {
            this(e, null);
        }

        public Node() {
            this(null, null);
        }

        @Override
        public String toString() {
            return e.toString();
        }
    }

    private Node dummyHead;
    private int size;
    
    public LinkList() {
        dummyHead = new Node();
        size = 0;
    }

    // 返回长度
    public int getSize() {
        return size;
    }

    // 是否为空
    public boolean isEmpty() {
        return size == 0;
    }

    // 添加到指定索引位置 
    public void add(int index, E e) {
        if (index < 0 || index > size) 
            throw new IllegalArgumentException("Index out of range.");
        Node pre = dummyHead;
        for (int i=0; i<index; i++) 
            pre = pre.next;
    
        Node node = new Node(e);
        node.next = pre.next;
        pre.next = node;
        size++;
    }

     // 添加到头部
     public void addFirst(E e) {
        add(0, e);
    }

    // 添加到末尾
    public void addLast(E e) {
        add(size, e);
    }

    // 查找 指定第几个
    public E get(int index) {
       if (index < 0 || index > size) 
            throw new IllegalArgumentException("Index out of range.");
        Node cur = dummyHead.next;
        for (int i=0; i<index; i++) {
            cur = cur.next;
        }
        return cur.e;
    }

    // 获取第一个
    public E getFirst() {
        return get(0);
    }

    // 获取最后一个
    public E getLast() {
        return get(size-1);
    }

    // 设置
    public void set(int index, E e) {
        if (index < 0 || index > size) 
            throw new IllegalArgumentException("Index out of range.");
        Node cur = dummyHead.next;
        for (int i=0; i<size; i++) {
            cur = cur.next;
        }
        cur.e = e;
    }

    // 查找元素
    public boolean find(E e) {
        Node cur = dummyHead.next;
        while(cur != null) {
            if (cur.e.equals(e))
                return true;
            cur = cur.next;
        }
        return false;
    }

    // 删除元素
    public E del(int index) {
        if (index < 0 || index > size) 
            throw new IllegalArgumentException("Index out of range.");
        Node pre = dummyHead;
        for (int i=0; i<index; i++) {
            pre = pre.next;
        }
        Node delNode = pre.next;
        pre.next = delNode.next;
        delNode.next = null;
        return delNode.e;
    }

    // 删除元素
    public void removeElement(E e) {
        Node prev = dummyHead;
        while(prev.next != null) {
            if (prev.next.e.equals(e)) 
                break;
            prev = prev.next;
        }

        if (prev.next != null) {
            Node tmpNode = prev.next;
            prev.next = tmpNode.next;
            tmpNode.next = null;
            size--;
        }
    }

    // 删除第一个元素
    public E delFirst() {
        return del(0);
    }

    // 删除最后一个元素
    public E delLast() {
        return del(size);
    }

    @Override
    public String toString() {
        StringBuilder res = new StringBuilder();
        Node cur = dummyHead.next;
        while(cur != null) {
            res.append(cur + "->");
            cur = cur.next;
        }
        res.append("NULL");
        return res.toString();
    }
}

File: golang/LinkList/linklist.go
package main

import (
	"fmt"
	"strings"
)

// LinkList 链表实现
// 使用虚拟头节点（dummy head）来简化操作
type LinkList[E any] struct {
	dummyHead *Node[E] // 虚拟头节点
	size      int      // 链表大小
}

// Node 链表节点
type Node[E any] struct {
	e    E        // 节点数据
	next *Node[E] // 下一个节点指针
}

// NewLinkList 创建一个新的链表
func NewLinkList[E any]() *LinkList[E] {
	return &LinkList[E]{
		dummyHead: &Node[E]{}, // 创建虚拟头节点
		size:      0,
	}
}

// newNode 创建一个新的节点
func newNode[E any](e E) *Node[E] {
	return &Node[E]{
		e:    e,
		next: nil,
	}
}

// GetSize 返回链表长度
func (l *LinkList[E]) GetSize() int {
	return l.size
}

// IsEmpty 判断链表是否为空
func (l *LinkList[E]) IsEmpty() bool {
	return l.size == 0
}

// Add 在指定索引位置添加元素
// index: 要插入的位置（0-based）
// e: 要插入的元素
func (l *LinkList[E]) Add(index int, e E) {
	if index < 0 || index > l.size {
		panic("Index out of range.")
	}

	// 找到插入位置的前一个节点
	pre := l.dummyHead
	for i := 0; i < index; i++ {
		pre = pre.next
	}

	// 创建新节点并插入
	node := newNode(e)
	node.next = pre.next
	pre.next = node
	l.size++
}

// AddFirst 在链表头部添加元素
func (l *LinkList[E]) AddFirst(e E) {
	l.Add(0, e)
}

// AddLast 在链表尾部添加元素
func (l *LinkList[E]) AddLast(e E) {
	l.Add(l.size, e)
}

// Get 获取指定索引位置的元素
func (l *LinkList[E]) Get(index int) E {
	if index < 0 || index >= l.size {
		panic("Index out of range.")
	}

	cur := l.dummyHead.next
	for i := 0; i < index; i++ {
		cur = cur.next
	}
	return cur.e
}

// GetFirst 获取第一个元素
func (l *LinkList[E]) GetFirst() E {
	return l.Get(0)
}

// GetLast 获取最后一个元素
func (l *LinkList[E]) GetLast() E {
	return l.Get(l.size - 1)
}

// Set 设置指定索引位置的元素
func (l *LinkList[E]) Set(index int, e E) {
	if index < 0 || index >= l.size {
		panic("Index out of range.")
	}

	cur := l.dummyHead.next
	for i := 0; i < index; i++ {
		cur = cur.next
	}
	cur.e = e
}

// Find 查找元素是否在链表中
// 注意：这里使用类型断言来比较元素，对于复杂类型需要实现自定义比较方法
func (l *LinkList[E]) Find(e E) bool {
	cur := l.dummyHead.next
	for cur != nil {
		// 使用类型断言进行相等性比较
		if l.equal(cur.e, e) {
			return true
		}
		cur = cur.next
	}
	return false
}

// Del 删除指定索引位置的元素
func (l *LinkList[E]) Del(index int) E {
	if index < 0 || index >= l.size {
		panic("Index out of range.")
	}

	// 找到要删除节点的前一个节点
	pre := l.dummyHead
	for i := 0; i < index; i++ {
		pre = pre.next
	}

	// 删除节点
	delNode := pre.next
	pre.next = delNode.next
	delNode.next = nil
	l.size--

	return delNode.e
}

// DelFirst 删除第一个元素
func (l *LinkList[E]) DelFirst() E {
	return l.Del(0)
}

// DelLast 删除最后一个元素
func (l *LinkList[E]) DelLast() E {
	return l.Del(l.size - 1)
}

// Remove 删除指定元素（如果存在）
func (l *LinkList[E]) Remove(e E) {
	// 找到要删除元素的前一个节点
	pre := l.dummyHead
	for pre.next != nil {
		if l.equal(pre.next.e, e) {
			// 删除节点
			delNode := pre.next
			pre.next = delNode.next
			delNode.next = nil
			l.size--
			return
		}
		pre = pre.next
	}
}

// String 返回链表的字符串表示
func (l *LinkList[E]) String() string {
	var res strings.Builder
	cur := l.dummyHead.next
	for cur != nil {
		res.WriteString(fmt.Sprintf("%v->", cur.e))
		cur = cur.next
	}
	res.WriteString("NULL")
	return res.String()
}

// equal 比较两个元素是否相等
// 这里使用类型断言，对于复杂类型需要实现自定义比较方法
func (l *LinkList[E]) equal(a, b E) bool {
	return fmt.Sprintf("%v", a) == fmt.Sprintf("%v", b)
}

// ToSlice 将链表转换为切片
func (l *LinkList[E]) ToSlice() []E {
	result := make([]E, l.size)
	cur := l.dummyHead.next
	for i := 0; i < l.size; i++ {
		result[i] = cur.e
		cur = cur.next
	}
	return result
}

// Clear 清空链表
func (l *LinkList[E]) Clear() {
	l.dummyHead.next = nil
	l.size = 0
}

// Reverse 反转链表
func (l *LinkList[E]) Reverse() {
	if l.size <= 1 {
		return
	}

	var prev *Node[E] = nil
	cur := l.dummyHead.next
	var next *Node[E]

	for cur != nil {
		next = cur.next
		cur.next = prev
		prev = cur
		cur = next
	}

	l.dummyHead.next = prev
}

// HasCycle 检测链表是否有环
func (l *LinkList[E]) HasCycle() bool {
	if l.size <= 1 {
		return false
	}

	slow := l.dummyHead.next
	fast := l.dummyHead.next.next

	for fast != nil && fast.next != nil {
		if slow == fast {
			return true
		}
		slow = slow.next
		fast = fast.next.next
	}

	return false
}
